### ISSUE TITLE
Shrink causing InvalidCastException

### ISSUE THREAD

**Issue by j4m3z0r** [mbdavid/LiteDB#1083]
Hi there, I just started using LiteDB for an embedded project. In order to ensure that the size of the database doesn't balloon over time, I set things up to call Shrink during initialization of my database-powered class. This shrink operation is triggering an InvalidCastException. Here's the full stacktrace:

```
Unhandled Exception:
System.InvalidCastException: Specified cast is not valid.
  at (wrapper castclass) System.Object.__castclass_with_cache(object,intptr,intptr)
  at LiteDB.PageService.GetPage[T] (System.UInt32 pageID) [0x0005a] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.IndexService.GetNode (LiteDB.PageAddress address) [0x0000b] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.IndexService+<FindAll>d__13.MoveNext () [0x00064] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LinqExtensions+<>c__DisplayClass2_0`2+<<DistinctBy>g___|0>d[TSource,TKey].MoveNext () [0x0008e] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.QueryCursor.Fetch (LiteDB.TransactionService trans, LiteDB.DataService data, LiteDB.BsonReader bsonReader) [0x00010] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteEngine+<Find>d__9.MoveNext () [0x00205] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LinqExtensions+<YieldBatchElements>d__1`1[T].MoveNext () [0x0008b] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteEngine+<>c__DisplayClass22_0.<Insert>b__0 (LiteDB.CollectionPage col) [0x0003f] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteEngine.Transaction[T] (System.String collection, System.Boolean addIfNotExists, System.Func`2[T,TResult] action) [0x0004e] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteEngine.Insert (System.String collection, System.Collections.Generic.IEnumerable`1[T] docs, LiteDB.BsonType autoId) [0x00041] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteEngine.InsertBulk (System.String collection, System.Collections.Generic.IEnumerable`1[T] docs, System.Int32 batchSize, LiteDB.BsonType autoId) [0x00051] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteEngine.Shrink (System.String password, LiteDB.IDiskService tempDisk) [0x00104] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteDatabase.Shrink (System.String password) [0x00032] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteDatabase.Shrink () [0x00013] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at stream.UploadQueue..ctor (System.String stateDirectory) [0x0003c] in <e7965cbc2917400c8a077c965e3fb213>:0
  at stream.Program.Main (System.String[] args) [0x0004f] in <e7965cbc2917400c8a077c965e3fb213>:0
```

This is using the packages that are currently on Nuget. The pertinent code is pretty much as you'd expect, but here it is nonetheless:

```
        private string _stateDirectory;
        private LiteDatabase _db;
        private LiteCollection<UploadJob> _uploadJobs;

        public UploadQueue(string stateDirectory)
        {
            _stateDirectory = Path.GetFullPath(stateDirectory);
            var dbName = Path.Combine(_stateDirectory, "state.db");
            _db = new LiteDatabase(dbName);
            _db.Shrink();
            _uploadJobs = _db.GetCollection<UploadJob>("uploadJobs");
            _uploadJobs.EnsureIndex(x => x.EnqueueTime);
            _uploadJobs.EnsureIndex(x => x.HashedName);
        }
```

**Comment by j4m3z0r** [comment]
Also of note, I just discovered that it has created hundreds of `state-temp-<n>.db` files. My process is managed by [supervisord](http://supervisord.org/), so it keeps restarting every time this exception occurs, which probably explains them. Nonetheless, I'd argue that this is probably undesirable -- is there another option here? Perhaps writing the temp databases to /tmp so that they're deleted on reboot?

**Comment by j4m3z0r** [comment]
The plot thickens: if I remove the `Shrink` call, I get the InvalidCastException when trying to open a file using `FileStorage`:

```
[syncer] syncError occurred in upload: System.AggregateException: One or more errors occurred. ---> System.InvalidCastException: Specified cast is not valid.
  at (wrapper castclass) System.Object.__castclass_with_cache(object,intptr,intptr)
  at LiteDB.PageService.GetPage[T] (System.UInt32 pageID) [0x0005a] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.IndexService.GetNode (LiteDB.PageAddress address) [0x0000b] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.IndexService.Find (LiteDB.CollectionIndex index, LiteDB.BsonValue value, System.Boolean sibling, System.Int32 order) [0x00032] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.QueryEquals+<ExecuteIndex>d__2.MoveNext () [0x00029] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LinqExtensions+<>c__DisplayClass2_0`2+<<DistinctBy>g___|0>d[TSource,TKey].MoveNext () [0x0008e] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.QueryCursor.Fetch (LiteDB.TransactionService trans, LiteDB.DataService data, LiteDB.BsonReader bsonReader) [0x00010] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteEngine+<Find>d__9.MoveNext () [0x0011a] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Boolean& found) [0x00045] in <7aa7d6192efc4d31a9896d7a49ef2104>:0
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00000] in <7aa7d6192efc4d31a9896d7a49ef2104>:0
  at LiteDB.LiteFileStream.GetChunkData (System.Int32 index) [0x00036] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteFileStream..ctor (LiteDB.LiteEngine engine, LiteDB.LiteFileInfo file, System.IO.FileAccess mode) [0x0001f] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at (wrapper remoting-invoke-with-check) LiteDB.LiteFileStream..ctor(LiteDB.LiteEngine,LiteDB.LiteFileInfo,System.IO.FileAccess)
  at LiteDB.LiteFileInfo.OpenRead () [0x00000] in <0e7fe05a592849968cfe23dfc2a26585>:0
  at LiteDB.LiteStorage.OpenRead (System.String id) [0x00007] in <0e7fe05a592849968cfe23dfc2a26585>:0
```

So I'm guessing that this is probably what caused the problem with `Shrink` in the first place.

**Comment by mbdavid** [comment]
Hi @j4m3z0r, for some reason your datafile has an invalid page, and shrink method needs read all pages (and then found this broken page). You can try call "Recovery" static method from LiteEngine to restore this page.
This was an undetected problem that occurs in some cases (maybe concurrent access, or different environment). If you could share how is your application uses LiteDB this can helps me to improve and detect this cases.

**Comment by j4m3z0r** [comment]
Hey, thanks for the quick reply.

The application here is super simple: I’m maintaining a queue of files to upload to a server. I’m writing those UploadJob objects to store basic metadata and writing the file data itself to the FileStorage store. There’s one writer and one reader who both access the dB via that _db reference from two different threads (I thought this was safe?). However, this is not a situation with high transaction rates: enqueueing writes and dequeueing reads happen every few minutes.

The one thing I can think of is that this is in an embedded, battery powered device. It does have situations where the power is cut without a clean shutdown. I thought that the journaling system you have in place would handle this, but it’s the one area I can think of where the access pattern is slightly unusual.

Is there anything I can do differently to avoid this kind of corruption? And should I call Recovery on each startup to fix any potential DB problems before trying to access the data? I thought it would be good to sync the file system after each write, but I couldn’t find an easy way to do that in C# (on Linux, specifically). My transaction rate is low enough, and the data important enough that this would make sense if there was a way to do it.

Given that the issue looked to be with FileStore, I’ve rolled back to storing the file data directly on the file system. No idea if this will improve matters over the long term, but it hasn’t broken yet. :)

**Comment by j4m3z0r** [comment]
Oh, and the environment: this is running on a NanoPi Neo Core 2 LTS (catchy, no?) running Armbian’s Ubuntu 18.04 build against Xamarin’s official ARM64 packages.

**Comment by j4m3z0r** [comment]
By the way, I did keep a copy of the corrupt database and could share it with you if it would help. I’d rather not post it publicly, but can send you a Google Drive link or similar if it would be useful.

**Comment by RussellHaley** [comment]
> Hey, thanks for the quick reply.
> 
> The application here is super simple: I’m maintaining a queue of files to upload to a server. I’m writing those UploadJob objects to store basic metadata and writing the file data itself to the FileStorage store. There’s one writer and one reader who both access the dB via that _db reference from two different threads (I thought this was safe?). However, this is not a situation with high transaction rates: enqueueing writes and dequeueing reads happen every few minutes.
> 
> The one thing I can think of is that this is in an embedded, battery powered device. It does have situations where the power is cut without a clean shutdown. I thought that the journaling system you have in place would handle this, but it’s the one area I can think of where the access pattern is slightly unusual.
> 
> Is there anything I can do differently to avoid this kind of corruption? And should I call Recovery on each startup to fix any potential DB problems before trying to access the data? I thought it would be good to sync the file system after each write, but I couldn’t find an easy way to do that in C# (on Linux, specifically). My transaction rate is low enough, and the data important enough that this would make sense if there was a way to do it.
> 
> Given that the issue looked to be with FileStore, I’ve rolled back to storing the file data directly on the file system. No idea if this will improve matters over the long term, but it hasn’t broken yet. :)

Hi, I'm just a curious interloper: If your overall read/write is so low, is there any reason you don't open the file only when you need it and close the file when complete?

**Comment by j4m3z0r** [comment]
Hey, good thought. I may try that. I didn’t do it initially because I’m more suspicious of filesystem locks than I am of mutexes, esp on non-standard kernels (the Armbian kernel is marked as experimental for this board, though seems better than the “official” image from the board manufacturer in that it doesn’t spontaneously crash with high load). But that might be an outdated prejudice at this point.

That said, this still wouldn’t guarantee that the data gets flushed to disk, right?

**Comment by RussellHaley** [comment]
> Hey, good thought. I may try that. I didn’t do it initially because I’m more suspicious of filesystem locks than I am of mutexes, esp on non-standard kernels (the Armbian kernel is marked as experimental for this board, though seems better than the “official” image from the board manufacturer in that it doesn’t spontaneously crash with high load). But that might be an outdated prejudice at this point.
> 
> That said, this still wouldn’t guarantee that the data gets flushed to disk, right?
I would expect that closing the file would flush it to the OS, but Linux write caching would still be a 'concern'. That said, with the level of throughput you're talking about, your devices filesystem would need to be under really heavy load to be an issue, and someone would need to yank the power at a very exact point in time. Your application would need to have captured/manipulated the data and files, but fail to flush to disk. If your using EXT3/4 then the journaling would likely take care of this last concern.

I'd say you're pretty safe, especially if you only write the database record *after* you've uploaded/downloaded (or otherwise manipulated) the target files.

**Comment by j4m3z0r** [comment]
Hey guys, quick follow-up here. What's the best way to ensure that my database remains consistent? Is the corruption of my database the expected kind, where a recovery operation is just cleaning out a journal or similar? Or is this the unexpected kind, where the contents of the database should be considered inconsistent?

If the latter, is there anything I can do to prevent the database reaching this state? Or do I need to just accept that the database may become corrupted sometimes?

(I realize that disk errors and such may cause irrecoverable inconsistencies, but I don't think that was what happened here)

**Comment by j4m3z0r** [comment]
@mbdavid it looks like the Recovery() method was added after the last release on nuget. Any chance you could do a new release with that change along with any other bugfixes since 4.1.4? Thanks!

**Comment by matheussige** [comment]
@j4m3z0r just so you can test it, I cloned this repo and ran LiteDB.Demo after setting the path to my corrupted db file and tested recovery. However, it didn't work. Hope it does for you!

**Comment by mbdavid** [comment]
Hi @j4m3z0r @matheussige - I will review all shrink method to check any problem. Also, I will add recovery method in shell tool.

**Comment by matheussige** [comment]
Thanks @mbdavid for looking into it. My issue is not related to the Shrink method (just "ordinary" database corruption on an encrypted db), I just ended up here and found out about the Recovery method and tried it out.

**Comment by mbdavid** [comment]
But @matheussige, did you used, in any time, Shrink method in your database before get corrupted?

**Comment by matheussige** [comment]
@mbdavid I did not. I have 2 Pages on my db where I keep some logs, and usually at least one of them gets corrupted often on our customer's machines. I'm trying to figure out what could be causing that.

**Comment by ooikengsiang** [comment]
I too face similar problem about LiteDB got corrupted. Not sure how my user did it. Maybe due to crash. I'm using LiteDB version 4.1.4, no Shrink function is called.
Dropping collection also becoming a problem too due to drop scan through all indexes and drop them. 
```
PageService.GetPage[T] (System.UInt32 pageID)
System.InvalidCastException: Specified cast is not valid.
(wrapper castclass) System.Object.__castclass_with_cache(object,intptr,intptr)
LiteDB    PageService.GetPage[T] (System.UInt32 pageID)
LiteDB    IndexService.GetNode (LiteDB.PageAddress address)
LiteDB    IndexService+<FindAll>d__13.MoveNext ()
LiteDB    CollectionService.Drop (LiteDB.CollectionPage col)
LiteDB    LiteEngine+<>c__DisplayClass5_0.<DropCollection>b__0 (LiteDB.CollectionPage col)
LiteDB    LiteEngine.Transaction[T] (System.String collection, System.Boolean addIfNotExists, System.Func`2[T,TResult] action)
LiteDB    LiteEngine.DropCollection (System.String collection)
LiteDB    LiteDatabase.DropCollection (System.String name)
```

Anyone have any idea to workaround the issue if the database happen to be in corrupted state?

**Comment by ooikengsiang** [comment]
Manage to duplicate it once after multiple time of trying to kill the application while LiteDB is running.
It seem like IndexService.FindAll is looking for all IndexPage for given collection, but one of them returned as DataPage instead of expected IndexPage thus the invalid cast exception.
Maybe the linked list of page got mess up half way when LiteDB try to update the data?

**Comment by lbnascimento** [comment]
Hi! With the objective of organizing our issues, we are closing old unsolved issues. Please check the latest version of LiteDB and open a new issue if your problem/question/suggestion still applies. Thanks!

**Comment by omfgicbf** [comment]
@mbdavid why are all these tickets being closed with no resolution when you said: "v4 will be maintained only for bugs and critical updates". Why push people with established projects to move to a whole new, not backwards compatible version that's only just come out of RC in the last month?

This was a legitimate issue that was (probably) resolved with a recent patch. Summarily closing it whilst providing no details leaves people depending on v4 in the dark.

**Comment by mbdavid** [comment]
Hi @omfgicbf, v4 will be publish with new patchs and reviews. I just had no time yet for this now. I'm giving more hard work to deliver a better v5. As soon as possible i'm backing to v4 to deliver a new patch fix.

**Comment by omfgicbf** [comment]
Hi @mbdavid, I appreciate your reply. The issue is not the timeline for a v4 update, nor the focus of effort on v5; both are completely understandable. The point was related to @lbnascimento closing outstanding v4 tickets with:-

> Please check the latest version of LiteDB and open a new issue if your problem/question/suggestion still applies

The above statement contradicts your position that v4 will continue to be supported for bugs and critical updates.
